### ISSUE TITLE
[LifetimeSafety] False-negative with dereferencing std::optional of libc++

### ISSUE THREAD

**Issue by usx95** [llvm/llvm-project#188832]
```cpp
#include <optional>
#include <string>

const std::string& test_opt_strings() {
    std::optional<std::string> strings_or;
    return *strings_or; // no error about return stack address.
}
```

https://godbolt.org/z/YYch9h3Te

This works fine without libc++ https://godbolt.org/z/MaY39G1Pd

This is a recent regression as this was ok before a weeks.

cc: @aeft as this might be related to your recent change to std::optional ? tnx!
